src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.damagesource;
+
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.function.BiFunction;
+import net.minecraft.world.damagesource.DamageSource;
+import net.neoforged.neoforge.event.entity.living.DamageBlockEvent;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * DamageContainer encapsulates aspects of the entity damage sequence so that
+ * relevant context related to damage dealt is accessible throughout the entire
+ * sequence.
+ * <h4>Note: certain values will be defaults until the stage in the sequence when they are set.</h4>
+ * <H3>Damage Sequence and uses</H3><ol>
+ * <li>Entity is hurt by a damage source</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.EntityInvulnerablityCheckEvent EntityInvulnerablityCheckEvent}
+ * fires and determines if the sequence can commence</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.living.EntityPreDamageEvent EntityPreDamageEvent} fires
+ * and gives access to this. Modifiers should be added here.</li>
+ * <li>{@link DamageBlockEvent} fires</li>
+ * <li>armor, enchantments, mob effect, and absorption modifiers are applied to the damage</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.living.IncomingDamageEvent IncomingDamageEvent} fires and
+ * provides final values for the preceding modifiers and the last chance to negate the damage but will not
+ * undo the preceding effects</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.living.DamageTakenEvent DamageTakenEvent} fires and provides
+ * an immutable perspective of what the entire sequence ended with. </li>
+ * </ol>
+ *
+ *
+ */
+public interface DamageContainer {
+    public enum Reduction {
+        /** Damage reduced from the effects of armor */
+        ARMOR,
+        /** Damage reduced from enchantments on armor */
+        ENCHANT,
+        /** Damage reduced from active mob effects */
+        MOBEFFECT,
+        /** Damage absorbed by absorption. */
+        ABSORPTION
+    }
+
+    /**
+     * @return the value passed into {@link net.minecraft.world.entity.LivingEntity#hurt(DamageSource, float)} before
+     *         any modifications are made.
+     */
+    float getOriginalDamage();
+
+    /**
+     * @return the current amount expected to be applied to the entity or used in subsequent damage calculations.
+     */
+    float getNewDamage();
+
+    /**
+     * @return The damage source for this damage sequence
+     */
+    DamageSource getSource();
+
+    /**
+     * Adds a callback modifier to the vanilla damage reductions. Each function will be performed in sequence
+     * on the vanilla value at the time the {@link Reduction} type is set by vanilla.
+     * <h4>Note: only the {@link net.neoforged.neoforge.event.entity.living.IncomingDamageEvent IncomingDamageEvent}
```

Comment:
Isn't this supposed to say `EntityPreDamageEvent`?

Reply:
yes.  fixed.